### PR TITLE
Fix converter leaks and undefined behavior

### PR DESCRIPTION
**Pull request by williamfgc**
## Proposed changes

Address remaining leaks and undefined behavior in converter tests. 
Move local container to a static scope as sanitizer finds out of scope use. 
Cache/thread-friendly, can't use `constexpr static std::array` due to lack of `inline constexpr` in C++14 see [here](https://stackoverflow.com/a/28846608/470441).
Add bounds check in valence calculation which creates a Heinsenbug due to undefined behavior in `converter_test_Bi_dirac`. 
converter tests should pass ASAN/LSAN checks after this PR.
Related to #3312

## What type(s) of changes does this code introduce?

- Bugfix
- Refactoring (no functional changes, no api changes)

### Does this introduce a breaking change?

- Yes

## What systems has this change been tested on?
Ubuntu 20.04, must pass CI

## Checklist

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/QMCTools/QMCGaussianParserBase.cpp [hunk 1]
```diff
@@ -475,8 +479,11 @@ xmlNodePtr QMCGaussianParserBase::createIonSet()
       hout.push(SpecieID.str().c_str(), true);
       int z          = static_cast<int>(ionSpecies(AtomicNumberIndex, i));
       double valence = ionSpecies(IonChargeIndex, i);
-      if (valence > CoreTable[z] && FixValence != true)
-        valence -= CoreTable[z];
+      if (z < gCoreTable.size())
+      {
+        if (valence > gCoreTable[z] && !FixValence)
+          valence -= gCoreTable[z];
+      }
```
Comment (ye-luo) [inline src/QMCTools/QMCGaussianParserBase.cpp:486]: Please add else and throw error.

Comment (williamfgc) [reply src/QMCTools/QMCGaussianParserBase.cpp:486]: @ye-luo a couple of things as it seems like a larger effort due to unintended consequences:
1. `converter_test_Bi_dirac` would trigger this error as `z` is going out of bounds (tests pass as gcc/clang are permissive and zero initialize memory for the most part). 
2. Throwing an exception generates leaks due to the skipped `xmlFree*` lines (mixing C++ and C)

Comment (prckent) [reply src/QMCTools/QMCGaussianParserBase.cpp:486]: The original code is simply wrong if someone runs an element heavier than Xe.  We should error abort for z too large and print a meaningful error message.

**Review by prckent — CHANGES_REQUESTED** [review]

**Comment by prckent** [comment]
I think that the original code is simply wrong if someone runs a too heavy element (>Xe). We run out of lookup table and should guard against it. e.g. Error abort and print a meaningful message.
(Deleted original comment by accident)

File: src/QMCTools/QMCGaussianParserBase.cpp (hunk 1 above)
Comment (camelto2) [reply src/QMCTools/QMCGaussianParserBase.cpp:486]: I think adding `FixValence = true;` to the DiracParser constructor would avoid this issue for the converter test. The valence info is read and set correctly in the ascii parsing of the dirac output. This is similar to the QPParser and LCAOHDFParser constructors

Comment (williamfgc) [reply src/QMCTools/QMCGaussianParserBase.cpp:486]: I'll try @camelto2 suggestion to have at least a leaky exception (but no test doing it). Thank you all.

**Comment by williamfgc** [comment]
This should be ready, the last commit prevents a leak for the original failing test and `FixValence=true` fixes it. I don't see converter leaks with this PR. Thanks!

**Review by camelto2 — APPROVED** [review]
LGTM

**Review by ye-luo — CHANGES_REQUESTED** [review]
Please add documentation for FixValence. Not obvious what it means. What actions are taken or not taken if set true/false.

**Comment by williamfgc** [comment]
@ye-luo good point, `QMCGaussianParserBase::FixValence` predates this PR. I tried finding it in doxygen [docs](https://docs.qmcpack.org/doxygen/doxy/d8/d2c/a05128.html#aa31023af5fb1ca1ca8fd9b676cefed0e) without luck. It seems like a `bool` flag only used in the lines of code this PR touches. That's another reason for requesting @camelto2 review on this as the `DiracParser` was recently added, I also see that `LCAOHDFParser` sets it to `true`. I'll defer to you and @prckent for the original motivation as I'll probably get it wrong :)

**Comment by ye-luo** [comment]
@camelto2 It will be super helpful if you can put the closest definition here and then @williamfgc can finalize this PR.

**Comment by camelto2** [comment]
> @camelto2 It will be super helpful if you can put the closest definition here and then @williamfgc can finalize this PR.

If I'm being honest, I'm not entirely sure why we need it at all. it is written to the xml as the 'valence' tag, but what is actually used internally by the code is 'charge'

What I think it it trying to do is add some definition of a distinction between the total number of electrons  vs. the  valence electrons in the outermost shell. In the example of say a BFD pseudopotential for O where we remove the He core, you have a total of 6 electrons with 2 removed. the 'charge' is read from the GAMESS output as 6, and the atomic number is 8. Then for GAMESS FixValence is false since it doesn't set it in the constructor, so it looks up in the table and sees that gCoreTable[z=8] = 2. So then it takes the actual number of electrons which is 6 and subtracts 2 to give 4. So nominally, the charge remains as 6,  but there are only 4 "valence" p electrons in this case while the 2s electrons would be considered semi-core. So that is what gets written to the structure.xml file. But to me it really isn't a meaningful distinction here, the only thing the code cares about the 'charge' element, which is correctly set by reading the GAMESS file. At least that is what I gather by quickly looking at code.

I think we could just drop it altogether and just leave 'valence' and 'charge' equal to one another for the ion species. That is what happens when FixValence is true. "valence" isn't even used by ParticleSet, so I don't even know why it is written to any of the xml files in the first place

**Comment by ye-luo** [comment]
Fixing/removing valence and charge input tags needs to be explored separately.

**Review by prckent — APPROVED** [review]
I agree with Ye's above comment. Explore removing/updating the supported tags in a separate issue/PR.

**Comment by ye-luo** [comment]
Test this please

**Pull request merged**